### ISSUE TITLE
AMD support and issues

### ISSUE THREAD

**Issue by dottodot** [balderdashy/sails.io.js#17]
I probably doing something wrong but I can't seem to get io.sails.url to work in my backbone app.

For example in my app.js I have to following

```
define(["backbone", "marionette", "sails.io"], function(Backbone, Marionette){
  io.sails.url = 'http://myapi.com:1337';
  var App = new Marionette.Application();
  return App;
});
```
However this isn't changing the url it's trying to connect to and following is reported in the console

```
WebSocket connection to 'ws://www.mysite.com/socket.io/1/websocket/%3C!DOCTYPE%20html%3E%3Chtml%20lang=%22en%22%3E%20%20%3Chead%3E%20%20%20%20%3Cmeta%20charset=%22utf-8%22%3E%20%20%20%20%3Cmeta%20name=%22viewport%22%20content=%22width=device-width,%20initial-scale=1.0%22%3E%20%20%20%20%3Ctitle%3EDatabase%3C/title%3E%20%20%20%20%20%20%20%20%3Clink%20href=%22/css/application.css%22%20rel=%22stylesheet%22%3E%20%20%20%20%3Clink%20rel=%22stylesheet%22%20type=%22text/css%22%20media=%22print%22%20href=%22/css/print.css%22%3E%20%20%3C/head%3E%20%20%3Cbody%3E%20%20%20%20%3Cdiv%20id=%22wrapper%22%3E%3C/div%3E%20%20%20%20%3Cdiv%20id=%22modal%22%20class=%22modal%20fade%22%20tabindex=%22-2%22%20role=%22dialog%22%20aria-hidden=%22true%22%3E%3C/div%3E%20%20%20%20%3Cscript%20async%20data-main=%22/js/require_main%22%20src=%22/js/components/requirejs/require.js%22%3E%3C/script%3E%20%20%3C/body%3E%3C/html%3E?__sails_io_sdk_version=0.10.0&__sails_io_sdk_platform=browser&__sails_io_sdk_language=javascript' failed: WebSocket is closed before the connection is established. 
```
followed several second latter by
```
GET http://www.mysite.com/socket.io/1/jsonp-polling/%3C!DOCTYPE%20html%3E%3Cht…dk_platform=browser&__sails_io_sdk_language=javascript&t=1406889826287&i=0 404 (Not Found) 
```

**Comment by hardchor** [comment]
Does anyone have a working shim config for sails.io.js in the meantime? It seems to be looking for "io" being in the global namespace, which won't be present when socket.io is loaded via requirejs.

**Comment by josser** [comment]
@hardchor i'm having same issue, did you find any ways to solve it?

**Comment by hardchor** [comment]
@josser Unfortunately not. I gave up and dumped it in the head for now.

**Comment by skainth** [comment]
I'm also struggling with getting sails.io work with requirejs. Any updates?

**Comment by sgress454** [comment]
@skainth it hasn't been a real priority, since you can work around it by simply adding `sails.io.js` separately before everything else.  If someone who's using requirejs a lot wants to work this out and submit a PR, we'd be happy to merge it in!

**Comment by mmacaula** [comment]
I'm having similar issues including sails.io with a webpack module bundler as well.  Workaround (putting in head) does work though.

**Comment by daedalus28** [comment]
I've patched the library to behave like the node version if it's running under AMD and will be opening a merge request in a few minutes. Since this library doesn't require socket io directly but instead uses constructor injection, there was no need to go full [UMD](https://github.com/umdjs/umd).

Just replace this at the end:
```js
  // Add CommonJS support to allow this client SDK to be used from Node.js.
  if (typeof module === 'object' && typeof module.exports !== 'undefined') {
    module.exports = SailsIOClient;
    return SailsIOClient;
  }

  // Otherwise, try to instantiate the client:
  // In case you're wrapping the socket.io client to prevent pollution of the
  // global namespace, you can replace the global `io` with your own `io` here:
  return SailsIOClient();
```

with this:

```js
    // Add CommonJS support to allow this client SDK to be used from Node.js.
    if (typeof module === 'object' && typeof module.exports !== 'undefined') {
        module.exports = SailsIOClient;
        return SailsIOClient;
    }
    else if (typeof define === 'function' && define.amd) {
        // AMD. Register as an anonymous module.
        define([], function() {
            return SailsIOClient;
        });
    }
    else {
        // Otherwise, try to instantiate the client:
        // In case you're wrapping the socket.io client to prevent pollution of the
        // global namespace, you can replace the global `io` with your own `io` here:
        return SailsIOClient();
    }
```


This issue was critical for me because I'm currently working on a project that (at least for now) needs to connect to a sails 0.9 server and a 0.11 server on different domains. I needed multiple versions of the sails and socket.io clients on the same page. Adding a script tag didn't cut it and using the shim config proved to be challenging because of the autoconnect logic.



As a side note, I think the main challenge with the autoconnect setting is that you need to time it just right - it needs to be after the library loads but before one timeout cycle runs. Other libraries that have similar config settings (like [requirejs](http://requirejs.org/docs/api.html#config)) handle this by allowing config to be defined on a specific global object. This means we don't need to wait for the library to load to specific set something like io.sails.autoconnect (which would understandably be undefined prior to loading) because you can just check if it exists prior to blowing it away.

**Comment by sailsbot** [comment]
Thanks for posting, @dottodot.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
